lib/mayaUsd/render/vp2RenderDelegate/material.cpp

```diff
@@ -1199,6 +1212,72 @@ TfToken MayaDescriptorToToken(const MVertexBufferDescriptor& descriptor)
 
 } // anonymous namespace
 
+class HdVP2Material::TextureLoadingTask
+{
+public:
+    TextureLoadingTask(
+        HdVP2Material*     parent,
+        HdSceneDelegate*   sceneDelegate,
+        const std::string& path,
+        const VtValue&     fallbackValue)
+        : _parent(parent)
+        , _sceneDelegate(sceneDelegate)
+        , _path(path)
+        , _fallbackValue(fallbackValue)
+    {
+    }
+
+    ~TextureLoadingTask() = default;
+
+    void Load()
+    {
+        if (_started.load()) {
+            return;
+        }
+        _started = true;
+        // Push the texture loading on idle
+        MGlobal::executeTaskOnIdle(
+            [](void* data) {
+                auto* task = static_cast<HdVP2Material::TextureLoadingTask*>(data);
+                task->_LoadOnIdle();
+                // Once it is done, free the memory.
+                delete task;
+            },
+            this);
+    }
+
+    bool Terminate()
+    {
+        _terminated = true;
+        // Return the started state to caller, the caller will delete this object
+        // if this task has not started yet.
+        // We will not be able to delete this object within its method.
+        return !_started.load();
+    }
+
+private:
+    void _LoadOnIdle()
+    {
+        if (_terminated) {
+            return;
+        }
+        bool        isSRGB = false;
+        MFloatArray uvScaleOffset;
+        auto*       texture = _LoadTexture(_path, isSRGB, uvScaleOffset, _fallbackValue);
+        if (_terminated) {
+            return;
+        }
+        _parent->_UpdateLoadedTexture(_sceneDelegate, _path, texture, isSRGB, uvScaleOffset);
+    }
+
+    HdVP2Material*    _parent;
+    HdSceneDelegate*  _sceneDelegate;
+    const std::string _path;
+    const VtValue     _fallbackValue;
+    std::atomic_bool  _started { false };
+    std::atomic_bool  _terminated { false };
```

Comment:
I think that _terminated doesn't need to be atomic. During the _loadOnIdle call none of the other bits of code which set _terminated true can be executing.

Reply:
Updated as well, `_terminated` is now just a `bool`.